### ISSUE TITLE
Why is there an input prompt in the replay while serving requests?

### ISSUE THREAD

**Issue by ibnesayeed** [oduwsdl/ipwb#447]
This code block makes me nervous, unless we plan to hire a team of encryption key typists who's only job is to keep the server running in non-daemon mode and be ready to enter appropriate encryption keys on each request that hits an encrypted record. Also, the prompt asks for providing path of a key file, but the expectations seems to be the key itself.

https://github.com/oduwsdl/ipwb/blob/01830df4d68c9451f66dae4f59de57eade6866a6/ipwb/replay.py#L622-L630

**Comment by machawk1** [comment]
If the key is removed from a CDXJ, a user will need another way to decrypt the contents. This provides the mechanism for the scenario where one receives a CDXJ without the necessary means to decrypt it but then uses a separate key file to do so.

Is it very rudimentary, as we have no spent a lot of time on the ipwb encryption aspects until very recently.

**Comment by machawk1** [comment]
You are right about it weird that the request for a file is in the request handler though. This is due for some refactoring.

**Comment by ibnesayeed** [comment]
Having a supplementary key file is fine as long as it is supplied at the time of server invocation, but this code is actually asking for the key on the server prompt on each request made from the client. This is impractical.

**Comment by machawk1** [comment]
> as long as it is supplied at the time of server invocation

At some point we want the indexed contents to be more dynamic at runtime. However, I agree that per-HTTP-request key file requests is impractical.

Can you anticipate the scenario where the client is asked for the key? This may be more fitting and would allow one-off URI-M decryption.

**Comment by ibnesayeed** [comment]
> At some point we want the indexed contents to be more dynamic at runtime.

This really does not make a difference. If there is a single key being used for the entire collection (which is not a great option) that it can perhaps be supplied at the time of booting the replay as an argument. If however different records might have different keys (and some might not have any), but the main CDXJ file does not have them inline, they can be offloaded to a separate key index. Lookup field for the key index file can either be SURT+14DigitDatetime (this can be ambiguous) or the hash of the payload.

```
$ cat index.cdxj
com,example,)/ 20180207132547 {"locator": "urin:ipfs/0123456789...", "encryption": {"method": "external"}, "...": "..."}
com,example,)/bar 20180207172819 {"locator": "urin:ipfs/1029384756...", "...": "..."}
com,example,)/foo 20180207092618 {"locator": "urin:ipfs/9876543210...", "encryption": {"method": "external"}, "...": "..."}

$ cat key.cdxj
0123456789... {"method": "aes", "nonce": "blah", "key": "IWontTellYou"}
9876543210... {"method": "aes", "nonce": "baz", "key": "NoItIsASecret"}
```

This way., when the main index suggests that the content is encrypted but the encryption details can be found from an external source, they supplied keys file should be used for lookup with the corresponding payload hash. This key file can be supplied as an argument when the replay is started the say way an index file is supplied. Also, we may want to never put encryption details in the main index and always keep them in a separate file like this and simply add `"encryption": true` in the index to indicate that the key index should be looked up.

> Can you anticipate the scenario where the client is asked for the key? This may be more fitting and would allow one-off URI-M decryption.

This is again an impractical approach. Even if we somehow try to throw a prompt for the end user to input some keys (such as by initiation an HTTP BasicAuth dialogue), it is going to be frustrating when a single page has hundreds of resources which which those keys are to be asked one by one. Besides, this is not an information the end user should provide one by one. The server has the index, so it has the knowledge of the resource and it should handle any encryption in place. As far as authorization is concerned, that should be dealt in a separate layer between the client and the server.

**Comment by machawk1** [comment]
> they can be offloaded to a separate key index.

I had yet to consider this but it would be interesting to have it as an option as a method of providing keys from users to encrypted mementos.

> "encryption": true

We discussed this yesterday and the idea of separating out all of the encryption attributes into their own JSON object, potentially contained within the record.

How about if a client were offered the option to provide a key.cdxj file for interpreting the mementos available for replay?

**Comment by ibnesayeed** [comment]
> We discussed this yesterday and the idea of separating out all of the encryption attributes into their own JSON object, potentially contained within the record.

Keep that in mind I provided that example above which is essentially a vertical split of the index, replicating the concept of database normalization. There are many ways to envision this. There could be an encryption identifier that can be used as the lookup column of the key file and recorded as a reference in the index. Using `true` as a value is one way to indicate that the record is encrypted and encryption information can be found in a provided key. This will be needed to identify those encrypted records from plain ones if the index has both, and this approach is only desired when there is only one way of providing keys, i.e., using a key file. Alternatively, the current option to embed that detail in the index itself can continue to work (we can certainly wrap related attributes in a single object).

> How about if a client were offered the option to provide a key.cdxj file for interpreting the mementos available for replay?

If the replay server is intermediary and does not have access to keys directly, then I would not encourage this approach of a user uploading keys file to it for replay because: 1) now the server needs to maintain session for each user and keep those keys forever for every single user every connected, 2) the server makes sure there is isolation, so keys of one user are not being used by another, and 3) the server now has access to private information of the clients which it did not have before.

If for whatever reason the client is the only one who has access to keys and the replay server is nothing but a proxy to fetch stuff from IPFS network, then the best option is to store keys in the browser itself and decrypt on the client-side (somewhat like end-to-end encryption). ServiceWorker can be helpful in this case where you perform decryption operation right before serving it to the browser.

**Comment by machawk1** [comment]
I made some adjustments in cc43d7c (issue-447 branch) to consider encryption as a JSON object instead of separate encryption attributes (e.g., `encryption_key`) as separate attributes on the main object.

Also, there was an issue of using `pad()` in Py3 due to byte/str mixup. cc43d7c fixes this, too.

More logic is needed to account for the external file. Conditionals are pretty hardcoded but this is a start, @ibnesayeed.

**Comment by machawk1** [comment]
Ping @Hitendrasinh99 , who is currently exploring this.
